### ISSUE TITLE
Wrong AL0582 error > Temporary parameters are not recognized in interface definitions

### ISSUE THREAD

**Issue by fvet** [microsoft/AL#6156]
Related to #5675 

**Describe the bug**
When declaring an interface with a procedure, containing a **temporary** parameter, implementing the interface with the exact same signature in **another extension**, fails.

**To Reproduce**
Steps and to reproduce the behavior:
1. In extension A, create an interface with a procedure containing a temporary parameter

```
interface "NVT ITaskProgram"
{
   procedure AddDetailToTempTaskTreeBuffer(var TempNVTTaskTreeBuffer: Record "NVT Task Tree Buffer" temporary; TptRegister: Record "NVT Tpt Register"; NewIndentation: Integer)
}
```

2. Create an enum

```
enum 2002616 "NVT Task Program" implements "NVT ITaskProgram"
{
    Extensible = true;
...
}
```

3. Create an extension B with a dependency on extension A

4. Extend the enum in extension B

```
enumextension 2009405 "NVT Customs Program" extends "NVT Task Program"
{
    value(2009405; "Create Customs Declaration")
    {
        Caption = 'Create Customs Declaration';
        Implementation = "NVT ITaskProgram" = "NVT Customs FWD Link";
    }
}
```

5. Create a codeunit that implements the interface

```
Codeunit 2009409 "NVT Customs FWD Link" implements "NVT ITaskProgram"
{
    procedure AddDetailToTempTaskTreeBuffer(var TempNVTTaskTreeBuffer: Record "NVT Task Tree Buffer" temporary; TptRegister: Record "NVT Tpt Register"; NewIndentation: Integer)
    begin

    end;
}
```

The alc compiler fails with AL0582 message: "'NVT Customs FWD Link' does not implement the interface member 'NVT ITaskProgram.AddDetailToTempTaskTreeBuffer'."

Remove the temporary from the parameter fixes the issue, although it is NOT correct.

```
Codeunit 2009409 "NVT Customs FWD Link" implements "NVT ITaskProgram"
{
    procedure AddDetailToTempTaskTreeBuffer(var TempNVTTaskTreeBuffer: Record "NVT Task Tree Buffer"; TptRegister: Record "NVT Tpt Register"; NewIndentation: Integer)
    begin

    end;
}
```

This problem does not occur when implementing the interface in a same extension. Temporary is properly accepted.

**Expected behavior**

The compiler should accept the temporary indication on procedure arguments and respect the interface implementation accross different extensions

**Screenshots**

Error, although interface is implemented correct.

![image](https://user-images.githubusercontent.com/25268332/93849926-6149f500-fcad-11ea-9b9c-b29c87893f7e.png)

No error, although interface is implemented wrong.

![image](https://user-images.githubusercontent.com/25268332/93850016-8b9bb280-fcad-11ea-9514-0e3ed89db4f0.png)

**5. Versions:**

 - AL Language: 

![image](https://user-images.githubusercontent.com/25268332/93850086-abcb7180-fcad-11ea-9e31-31ad1a07772e.png)

**Comment by fvet** [comment]
@thloke Did something change recently? In VS Code, using the latest compiler, I'm still forced to remove the 'temporary' parameter to avoid compilation errors.

![image](https://user-images.githubusercontent.com/25268332/94427964-3b7e8d80-0190-11eb-996c-a6c5ddf59173.png)

However, when **publishing** on an insider container (Artifact Url: [https://bcinsider.azureedge.net/sandbox/**17.1.17069.0**/be?xxxxx), I now get the same error upon publishing the app

```
2020-09-27T01:12:45.4989409Z  * App File = [C:\Run\DevOps\Artifacts\Apps\Navitrans_Customs_16.3.0.4308_APP.app]
2020-09-27T01:12:45.4994111Z 
2020-09-27T01:12:45.5020220Z  * App.ID        = ca205d1e-c069-40d5-ac10-7feb3da9faf0
2020-09-27T01:12:45.5030925Z  * App.Name      = Customs
2020-09-27T01:12:45.5033190Z  * App.Publisher = Navitrans
2020-09-27T01:12:45.5039153Z  * App.Version   = 16.3.0.4308
2020-09-27T01:12:45.5316510Z WARNING: This license is not compatible with this version of Business Central.
2020-09-27T01:12:45.5316742Z  => Publishing App 'Customs v16.3.0.4308'
2020-09-27T01:12:45.5316948Z WARNING: This license is not compatible with this version of Business Central.
2020-09-27T01:12:45.7154687Z Extension compilation failed
2020-09-27T01:12:45.7156858Z src/Forwarding/COD%20Customs%20FWD%20Link.al(1,52): error AL0582: 'NVT Customs FWD Link' does not implement the interfa
2020-09-27T01:12:45.7157307Z ce member 'NVT ITaskProgram.AddDetailToTempTaskTreeBuffer'.
2020-09-27T01:12:45.7157525Z At line:155 char:17
2020-09-27T01:12:45.7157728Z + ...             $Result = Invoke-Command -Session $DockerSession -ScriptB ...
2020-09-27T01:12:45.7157967Z +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-09-27T01:12:45.7158255Z     + CategoryInfo          : InvalidOperation: (:) [Publish-NAVApp], InvalidOperationException
2020-09-27T01:12:45.7158592Z     + FullyQualifiedErrorId : MicrosoftDynamicsNavServer$BC/nav-systemapplication,Microsoft.Dynamics.Nav.Apps.Manageme 
2020-09-27T01:12:45.7158849Z    nt.Cmdlets.PublishNavApp
2020-09-27T01:12:45.7158957Z  
2020-09-27T01:12:45.7159028Z 
2020-09-27T01:12:45.7214326Z ##[error]Batch App-Publish failed: Extension compilation failed
2020-09-27T01:12:45.7220187Z src/Forwarding/COD%20Customs%20FWD%20Link.al(1,52): error AL0582: 'NVT Customs FWD Link' does not implement the interface member 'NVT ITaskProgram.AddDetailToTempTaskTreeBuffer'.
```

What's the workaround?

**Comment by thloke** [comment]
@fvet, just to clarify, you get the error on publishing with the interface implementation **without** the temporary modifier?

If so, that's something we'll need to investigate/fix. I'll raise this with the team today to get this prioritized.

**Comment by fvet** [comment]
Short answer: **yes.**

Long answer: I just tested again on a new container, even on an earlier version and we have the same issue too. All pipelines for different target are failing.

![image](https://user-images.githubusercontent.com/25268332/94687273-5be34f00-032c-11eb-92fc-edf1b6649ff5.png)

Interface in Extension 1

![image](https://user-images.githubusercontent.com/25268332/94687888-24c16d80-032d-11eb-975d-2af1f0ee965a.png)

Interface implementation in Extension 2

![image](https://user-images.githubusercontent.com/25268332/94687969-428ed280-032d-11eb-9a0e-b78bc9788587.png)

VSCode background compilation fails if temporary is set, so removed temporary.

However, VSCode fails when publishing this code (the compiler is fine without temporary, but publishing fails without temporary)

![image](https://user-images.githubusercontent.com/25268332/94688208-913c6c80-032d-11eb-8d46-94a030d5f11d.png)

**Comment by thloke** [comment]
@PooyaKharamesh - can you advise if this is the same issue that you fixed recently, and what the fix version was? It's strange that the compilation on VSCode and the compilation on the server has different behaviour.

**Comment by gvae** [comment]
Any news on this topic? @PooyaKharamesh

**Comment by fvet** [comment]
FYI, after reading https://vjeko.com/2012/11/01/some-tips-and-hints-about-temporary-tables/, we've changed the conflicting code for the function involved not to receive the 'temporary' indicator any longer. This is now no longer blocking for us ... nevertheless, it would be nice if the bug can still be fixed. (or introduce another check in AL not to accept any functions with 'var record temporary' arguments)

**Comment by PooyaKharamesh** [comment]
This issue is fixed and backported. The fix is available in BC 16.4 and higher versions. 

/Pooya
